SERVERLESS-354 Implement log shipping inside the go-proxy

markusthoemmes opened the pull request:
This is a first whack at an actual implementation. There are some open questions marked as TODO but I wanted to get this out of my local development to allow people to take a first look at things.

---

This is a first implementation of 3rd party log-shipping inside the go-proxy. If the respective env vars are set, stdout and stderr from the function process are intercepted and written to the service as configured.

This contains initial support for DataDog, Papertrail and Logtail. Only HTTPS transport is currently supported.

markusthoemmes committed:
SERVERLESS-354 Implement log shipping inside the go-proxy

This is a first implementation of 3rd party log-shipping inside the go-proxy. If the respective env vars are set, stdout and stderr from the function process are intercepted and written to the service as configured.

This contains initial support for DataDog, Papertrail and Logtail. Only HTTPS transport is currently supported.

jonfriesen commented on openwhisk/executor.go:
```diff
@@ -68,17 +87,151 @@ func NewExecutor(logout *os.File, logerr *os.File, command string, env map[strin
+	case "logtail":
+		proc.remoteLogger = &httpLogger{
+			url:     "https://in.logtail.com",
+			headers: map[string]string{"Authorization": fmt.Sprintf("Bearer %s", env["LOGTAIL_SOURCE_TOKEN"])},
```
do the `env["LOGTAIL_SOURCE_TOKEN"]`'s in this section need to be validated?

markusthoemmes replied:
Probably yes. I'll add some validation here and open a JIRA to potentially add validation to the controller at action creation time.

markusthoemmes replied:
Added verification code.

jonfriesen commented on openwhisk/executor.go:
```diff
@@ -68,17 +87,151 @@ func NewExecutor(logout *os.File, logerr *os.File, command string, env map[strin
+	// A chunky buffer in order to not block execution of the function from dealing with log lines.
+	proc.lines = make(chan LogLine, 128)
+
+	// TODO: Should we throw stdout and stderr into one bucket? We'd lose the ability to
+	// distinguish them but, I think, we'd gain the ability to ensure that the ordering is
+	// equal to how the lines are ordered in-code (if it mixes stdout and stderr).
```
in `consumeStream(...)` we wrap the log in a LogLine, could we combine the two, and indicate the stream within that metadata? Then users can filter for stdout/stderr?

markusthoemmes replied:
I'm not quite parsing the question tbh. If we combine the two, we lose the ability to distinguish stdout/stderr. However, anecdotally, the AP implementation doesn't pass on any metadata about stdout/stderr either, so that might be okay? I guess it's not usual to log to stdout and stderr at the same time for apps and stuff. It's more common in CLI tools, where sometimes the rule is to write results to stdout and **any** logs (not just errors) to stderr.

jonfriesen commented on openwhisk/executor.go:
```diff
@@ -68,17 +87,151 @@ func NewExecutor(logout *os.File, logerr *os.File, command string, env map[strin
+	}
+	proc.consumeGrp.Go(func() error { return consumeStream("stderr", stderr, proc.lines) })
+
+	// TODO: Support multiple destinations at once.
```
Should this be tied to a jira?

markusthoemmes replied:
Yep, will open one.

jonfriesen reviewed (approved):
over all LGMT. I'd like to see some of the TODO's be removed/turned into Jira tickets.

markusthoemmes committed:
Review adjustments

- Moved env vars into constants
- Added config verification
- A few test touchups

markusthoemmes commented:
Thanks @jonfriesen. Pushed the adjustments as commented.

markusthoemmes committed:
Fix test name

jonfriesen reviewed (approved):
lgtm :ship:

markusthoemmes committed:
Drop TODO

piyush0101 commented on openwhisk/executor.go:
```diff
@@ -68,17 +95,160 @@ func NewExecutor(logout *os.File, logerr *os.File, command string, env map[strin
+					// We've seen both sentinels. Stop consuming until the next request.
+					return lastErr
+				}
+				continue
```
do you need this `continue`? If we saw the sentinels, the code returns `lastErr`, else, it continues to write to the `remoteLogger`

markusthoemmes replied:
Yes because we don't want to write the sentinels to the remote. This `continue` is ran if we saw the stdout but not the stderr sentinel yet.

piyush0101 replied:
gotcha, makes sense.

piyush0101 commented on openwhisk/executor.go:
```diff
@@ -68,17 +95,160 @@ func NewExecutor(logout *os.File, logerr *os.File, command string, env map[strin
+			if err := proc.remoteLogger.Send(line); err != nil {
+				// We want to continue consuming all of the logs so we don't exit immediately
+				// but surface the error eventually.
+				lastErr = fmt.Errorf("failed to send log to remote location: %w", err)
```
what if there are multiple errors? Should we accumulate errors in something like a https://github.com/hashicorp/go-multierror and return?

markusthoemmes replied:
Can do, yeah.

markusthoemmes replied:
Actually, via the demo call, I'll log the error in-line and not return with an error from here. We'll log the error + the log line we tried to write so it ain't lost. Does that sound fair?

markusthoemmes replied:
Done!

piyush0101 commented on openwhisk/executor.go:
```diff
@@ -68,17 +95,160 @@ func NewExecutor(logout *os.File, logerr *os.File, command string, env map[strin
+
+	proc.logout.WriteString(remoteLogNotice)
+	proc.logout.WriteString("\n")
+	out, err := proc.roundtrip(in)
```
curious why is the `roundtrip` necessary?

markusthoemmes replied:
It's what's actually calling the function to make it run.

piyush0101 commented on openwhisk/executor.go:
```diff
@@ -68,17 +95,160 @@ func NewExecutor(logout *os.File, logerr *os.File, command string, env map[strin
+
+	// Wait for the streams to process completely.
+	if err := grp.Wait(); err != nil {
+		fmt.Fprintf(proc.logerr, "Failed to process logs: %v\n", err)
```
its probably good to return from here rather than continuing through the rest of the statements below

markusthoemmes replied:
Hmm, I see where you're going but we have to write the OutputGuards actually. Maybe I should put them into a `defer` and then exit early here.

markusthoemmes replied:
I have to correct: We actually don't want to return this error but the error from roundtrip everywhere. I'll do the defer and delete the flush, that'll come down to roughly the same thing then.

piyush0101 commented on openwhisk/executor.go:
```diff
@@ -68,17 +95,160 @@ func NewExecutor(logout *os.File, logerr *os.File, command string, env map[strin
+	}
+
+	if err := proc.remoteLogger.Flush(); err != nil {
+		fmt.Fprintf(proc.logerr, "Failed to flush logs to remote location: %v\n", err)
```
same comment as above.

piyush0101 replied:
I saw that `Flush` is not implemented, should we remove this from here, seems like unused code?

markusthoemmes replied:
Ack, will reintroduce it once it makes sense.

markusthoemmes committed:
Review changes

- Defer log line writing
- Collect all errors into multierr
- Drop Flush

markusthoemmes committed:
Golang 1.15 compatibility
